packages/vm/src/instance.rs

```diff
@@ -62,15 +62,22 @@ where
         options: InstanceOptions,
         memory_limit: Option<Size>,
     ) -> VmResult<Self> {
-        let module = compile(code, memory_limit)?;
-        Instance::from_module(&module, backend, options.gas_limit, options.print_debug)
+        let module = compile(code, memory_limit, &[])?;
+        Instance::from_module(
+            &module,
+            backend,
+            options.gas_limit,
+            options.print_debug,
+            None,
+        )
     }
 
     pub(crate) fn from_module(
         module: &Module,
         backend: Backend<A, S, Q>,
         gas_limit: u64,
         print_debug: bool,
+        extra_imports: Option<HashMap<&str, HashMap<&str, Function>>>,
```

Comment:
And empty map and None are the same thing, right? Do you use the Option to make this easy to call?

Reply:
Yes, basically. I feel weird about having to write `HashMap::new()` when calling this and not wanting to inject extra fns. Also slightly worried something's allocated that way, but maybe nothing is.

Reply:
Thanks for clarification. Makes sense.

By the way, Rust 1.56.0 seems to allow passing `[]` into an `Impl Into<HashMap<XYZ>>`: https://github.com/rust-lang/rust/commit/daf6f9951e25561cc97ae8715807b880f496da90

Reply:
An alternative could be a kind of Builder-ish pattern, i.e. `with_extra_imports()`.

Reply:
Or some good old mutable setter method.

I just wonder if we can change imports after this line was executed in `from_module`:

```
let wasmer_instance = Box::from(WasmerInstance::new(module, &import_obj).map_err(
```

Probably not.

Reply:
I considered a proper builder pattern for `Instance`s and even prototyped it somewhere, but then figured it's maybe a bit much for something that's internal and not a public API. It sure as hell would be nice to use though.

What could happen is we have a separate `InstanceBuilder` that first lets you set a bunch of options and then build an instance from them.

```rust
struct Instance {
    // the same structure we currently have
}

impl Instance {
    pub(crate) fn builder(module: &Module) -> InstanceBuilder {
        InstanceBuilder::from_module(module)
    }

    // probably keep from_code for backwards compatibility
}

/// exported through internals?
pub struct InstanceBuilder {
    gas_limit: u64,
    print_debug: bool,
    extra_imports: Option<HashMap<&str, Exports>>,
    // other options
}

impl InstanceBuilder {
    fn from_module(module: &Module) -> Self {
        // ...
    }

    pub fn gas_limit(&mut self, gas_limit: u64) -> &mut Self {
        // ...
    }

    pub fn import_fns(&mut self, namespace: impl Into<String>, import_fns: Exports) -> &mut Self {
        // ...
    }

    // ...more setters...

    fn build(&mut self) -> VmResult<Instance> {
        // the code that's currently in `Instance::from_module()`
    }
}
```

And now we can...

```rust
let exports = Exports::new();
// ...add functions to `exports`...

let instance = Instance::builder()
    .gas_limit(5)
    .print_debug(true)
    .import_fns("extra", exports)
    .build()?;
```